spec/fixtures/generate_external_certs.sh

```diff
@@ -0,0 +1,337 @@
+  # The altnames can be copied from the request potentially, by using the "copy_extensions = copy" option of openssl conf
+  cp $DIR_CONF/ca_$2.conf $DIR_CONF/site_$1.conf
+  if [ "$7" != "0" ]; then append_altnames    $7 $DIR_CONF/site_$1.conf; fi
+  if [ "$8" != "0" ]; then append_ip_altnames $8 $DIR_CONF/site_$1.conf; fi
+  
+  openssl ca -batch \
+    -config $DIR_CONF/site_$1.conf \
+    -extensions req_ext \
+    -out $DIR_CERTS/site_$1.crt \
+    -md $3 \
+    -startdate $START_DATE -enddate $END_DATE \
+    -infiles $CSR_DIR/site_$1.csr
+}
+
+function create_site_crt_selfsigned() { # $1 domain, $2 message_digest, $3 rsa_bits, $4 good/expired, $5 subject, $6 altnames, $7 ipaltnames
+  echo "Generating site key \"$1\""
+  openssl genrsa -out $DIR_KEYS/site_$1.key $3 2>/dev/null
+  
+  init_root_ca dummy "/CN=US/O=US/OU=US Unit"
+
+  FILE=/tmp/ca_gen/site_$1.req.conf
+  site_request_conf $FILE $3
+  if [ "$6" != "0" ]; then append_altnames    $6 $FILE; fi
+  if [ "$7" != "0" ]; then append_ip_altnames $7 $FILE; fi
+  
+  openssl rsa -in $DIR_KEYS/site_$1.key -out $DIR_KEYS/site_rsa_$1.key
+  openssl req -new -sha256 -key $DIR_KEYS/site_$1.key -subj "$5" -out $CSR_DIR/site_$1.csr -config $FILE
+
+  cp $DIR_CONF/ca_dummy.conf $DIR_CONF/site_$1.conf
+  
+  # Generate an extensions file; not extending default config since x509 command doesn't have a way to specify a
+  # full config file that I see.
+  #echo -e "[req_ext]\nsubjectAltName = @alt_names\n" >> conf/site_$1.conf
+  #echo -e "[alt_names]" >> conf/site_$1.conf
+  if [ "$6" != "0" ]; then append_altnames    $6 $DIR_CONF/site_$1.conf; fi
+  if [ "$7" != "0" ]; then append_ip_altnames $7 $DIR_CONF/site_$1.conf; fi
+  
+  gen_start_end $4
+  
+  #openssl x509 -req \
+  #  -in $CSR_DIR/site_$1.csr \
+  #  -$2 \
+  #  -extensions req_ext \
+  #  -out certs/site_$1.crt \
+  #  -signkey keys/site_rsa_$1.key \
+  #  -extfile conf/site_$1.conf
+  
+  openssl ca -batch \
+    -config $DIR_CONF/site_$1.conf \
+    -out $DIR_CERTS/site_$1.crt \
+    -extensions req_ext \
+    -selfsign \
+    -keyfile $DIR_KEYS/site_rsa_$1.key \
+    -md $2 \
+    -startdate $START_DATE -enddate $END_DATE \
+    -infiles $CSR_DIR/site_$1.csr
+}
+
+# Note that an altname is specified below. If req_ext is used, then this file cannot be used as is, because an empty
+# alt_names section is not allowed. It is done this way so the file can be copied and a list of altnames
+# be appended to the end of the copy.
+function generate_conf() { # $1 ca_name
+  cat >$DIR_CONF/ca_$1.conf << EOL
+[ca]
+default_ca = default
+[default]
+certs = $DIR_CADATA/$1/
+new_certs_dir = $DIR_CADATA/$1/db.certs
+database = $DIR_CADATA/$1/db.index
+serial = $DIR_CADATA/$1/db.serial
+certificate = $DIR_CERTS/ca_$1.crt
+private_key = $DIR_KEYS/ca_$1.key
+default_days = 365
+default_crl_days = 30
+default_md = sha256
+preserve = no
+RANDFILE = $DIR_CADATA/$1/db.random
```

Comment:
The RANDFILE option should actually not be specified here. It defaults to system random, and I've seen some problems where the file does not get created automatically in some cases.

Reply:
Dropped the RANDFILE line.